src/parser.ts

```diff
@@ -0,0 +1,39 @@
+interface Token {
+  type: string;
+  value?: string;
+}
+
+function* tokenizer(input : String) : Generator<Token, Token, void> {
+  const whitespace = /\s/;
+  const symbols = /[.*/_^[\]]/;
+  let cur = 0;
+
+  while (cur < input.length) {
+    if (symbols.test(input[cur])) {
+      yield { type: input[cur] };
+      cur++;
+    } else if (whitespace.test(input[cur])) {
+      const tok = { type: 'WHITESPACE', value: '' };
+      while (whitespace.test(input[cur])) {
+        tok.value = tok.value.concat(input[cur]);
+        cur++;
+      }
+      yield tok;
+    } else {
+      const tok = { type: 'WORD', value: '' };
+      while (
+        cur < input.length &&
+        !symbols.test(input[cur]) &&
+        !whitespace.test(input[cur])
+      ) {
+        tok.value = tok.value.concat(input[cur]);
+        cur++;
+      }
+      yield tok;
+    }
+  }
+
+  return { type: 'EOF' }
```

Comment:
I have this here ONLY so that the type checking of tests will pass, because the LAST token returned will be from return, and not from the last yield... (which is stupid).